File: LogFileHandler_src/LogFileHandler/LogFileManagers.vb
Imports System
Imports System.Collections.Generic
Imports System.Diagnostics
Imports System.Drawing
Imports System.IO
Imports System.Linq
Imports System.Security.AccessControl
Imports System.Text
Imports System.Threading
Imports System.Threading.Tasks

Public Class LogFileManagers
    Public Event OnFileAmount(ByVal fileAmount As Integer)
    Public Event OnFileAmountChanged(ByVal fileAmount As Integer)
    Public Event OnStatusReport(ByVal msg As String)
    Public Event OnFileCompleted(ByVal FullfileName As String)
    Public Event OnInternalError(ByVal ex As Exception)

    Private Shared GoodFilePath As String = Directory.GetCurrentDirectory() & "\GoodFiles\"
    Private Shared ErrorFilePath As String = Directory.GetCurrentDirectory() & "\ExceptionFiles\"
    Private Shared CSVPath As String = Directory.GetCurrentDirectory() & "\CSVFiles\"
    Private Shared logFilePath = Directory.GetCurrentDirectory() & "\LogFiles"
    Public Sub StartParserLogFiles()
        Try
            If Not Directory.Exists(GoodFilePath) Then
                Directory.CreateDirectory(GoodFilePath)
            End If
            If Not Directory.Exists(ErrorFilePath) Then
                Directory.CreateDirectory(ErrorFilePath)
            End If
            If Not Directory.Exists(CSVPath) Then
                Directory.CreateDirectory(CSVPath)
            End If
            If Not Directory.Exists(logFilePath) Then
                Directory.CreateDirectory(logFilePath)
            End If

            RaiseEvent OnStatusReport("Prepare Dummy log files")
            GenerateLogFiles()
            Dim files = Directory.GetFiles(logFilePath, "*.log")
            RaiseEvent OnFileAmount(files.Length)
            RaiseEvent OnStatusReport("Total [" & files.Length.ToString() & "] files are ready")

            ParallelExecution(files)
        Catch ex As Exception
            RaiseEvent OnInternalError(ex)
        End Try
    End Sub

    Private Sub ParallelExecution(ByVal files As String())
        Try
            Task.Run(Sub()
                         Dim stopwatch As Stopwatch = Stopwatch.StartNew()
                         Parallel.ForEach(files, New ParallelOptions With {
                             .MaxDegreeOfParallelism = Environment.ProcessorCount
                         }, Sub(currentFile)
                                Dim ExtractedResult = ExtractFileContent(ReadFileContent(currentFile))
                                StoreResultToCSVFile(ExtractedResult)
                                MoveFileToErrorFolder(currentFile)
                                RaiseEvent OnFileCompleted(String.Format("The file [{0}] is done", currentFile))
                                RaiseEvent OnFileAmountChanged(1)
                            End Sub)
                         RaiseEvent OnFileCompleted("Parallel Execution Time: " & stopwatch.ElapsedMilliseconds)
                         stopwatch.[Stop]()
                     End Sub)
        Catch e As Exception
            WriteDataToLocalFiles(e.Message)
            RaiseEvent OnInternalError(e)
        End Try
    End Sub

    Private Function ReadFileContent(ByVal sourceFile As String) As String
        Dim fileContent As String = String.Empty
        Try
            Using sr As StreamReader = New StreamReader(sourceFile, Encoding.Unicode)
                fileContent = sr.ReadToEnd()
            End Using

        Catch e As Exception
            WriteDataToLocalFiles(e.Message)
            RaiseEvent OnInternalError(e)
        End Try

        Return fileContent
    End Function

    Private Sub MoveFileToErrorFolder(ByVal sourceFile As String)
        Try
            Dim fileName As String = "_Err_" & Path.GetFileName(sourceFile)
            File.Move(sourceFile, ErrorFilePath & fileName)
        Catch e As Exception
            WriteDataToLocalFiles(e.Message)
            RaiseEvent OnInternalError(e)
        End Try
    End Sub

    Private Sub MoveFileToGoodFolder(ByVal sourceFile As String)
        Try
            File.Move(sourceFile, GoodFilePath)
        Catch e As Exception
            WriteDataToLocalFiles(e.Message)
            RaiseEvent OnInternalError(e)
        End Try
    End Sub

    Public Sub WriteDataToLocalFiles(ByVal strData As String)
        Try
            Dim path = Directory.GetCurrentDirectory()
            Dim folder As String = path & "\ApplicationLogs\"

            If Not Directory.Exists(folder) Then
                Directory.CreateDirectory(folder)
            End If

            Dim fileName As String = folder & String.Format("{0:yyyyMMddhh}", DateTime.Now) & ".log"

            Using fs = New FileStream(fileName, FileMode.OpenOrCreate, FileSystemRights.AppendData, FileShare.ReadWrite, 4096, FileOptions.Asynchronous)

                Using wr = New StreamWriter(fs)
                    wr.AutoFlush = True
                    wr.WriteLine(String.Format("{0:yyyy-MM-dd HH:mm:ss.ffff}", DateTime.Now))
                    wr.WriteLine(strData)
                End Using
            End Using

        Catch ex As Exception
            RaiseEvent OnInternalError(ex)
        End Try
    End Sub

    Private Sub StoreResultToCSVFile(ByVal strData As String)
        Try
            Dim fileName As String = CSVPath & String.Format("{0:yyyyMMddhh}", DateTime.Now) & ".csv"

            Using fs = New FileStream(fileName, FileMode.OpenOrCreate, FileSystemRights.AppendData, FileShare.ReadWrite, 4096, FileOptions.Asynchronous)

                Using wr = New StreamWriter(fs)
                    wr.AutoFlush = True
                    wr.WriteLine(strData)
                End Using
            End Using

        Catch ex As Exception
            WriteDataToLocalFiles(ex.Message)
            RaiseEvent OnInternalError(ex)
        End Try
    End Sub

    Public Sub GenerateLogFiles()
        Try
            Dim Qty As Integer = 10
            Parallel.[For](0, Qty, Sub(x)
                                       Dim destFile As String = logFilePath & "\" & String.Format("{0:yyyyMMddHHmmssffff}", DateTime.Now) & "_" & x.ToString() & ".log"
                                       Using fs As FileStream = File.Create(destFile)
                                           AddText(fs, "This is some text")
                                           AddText(fs, "This is some more text,")
                                           AddText(fs, vbCrLf & "and this is on a new line")
                                           AddText(fs, vbCrLf & vbCrLf & "The following is a subset of characters:" & vbCrLf)

                                           For i As Integer = 1 To 120 - 1
                                               AddText(fs, Convert.ToChar(i).ToString())
                                           Next
                                       End Using
                                   End Sub)
        Catch ex As Exception
            WriteDataToLocalFiles(ex.Message)
            RaiseEvent OnInternalError(ex)
        End Try
    End Sub

    Private Shared Sub AddText(ByVal fs As FileStream, ByVal value As String)
        Dim info As Byte() = New UTF8Encoding(True).GetBytes(value)
        fs.Write(info, 0, info.Length)
    End Sub

    Public Shared Function ExtractFileContent(ByVal fileContent As String) As String
        Return "a,b,c,d"
    End Function
End Class

File: LogFileHandler(vb)/LogFileHandler/LogFileManagers.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Security.AccessControl;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

public static class LogFileManagers
{
    public delegate void OnFileAmountHandler(int fileAmount);
    public static event OnFileAmountHandler OnFileAmount;

    public delegate void OnFileAmountChangedHandler(int fileAmount);
    public static event OnFileAmountChangedHandler OnFileAmountChanged;

    public delegate void StatusReportHandler(string msg);
    public static event StatusReportHandler OnStatusReport;

    public delegate void FileCompletedHandler(string FullfileName);
    public static event FileCompletedHandler OnFileCompleted;

    public delegate void ErrorHandler(Exception ex);
    public static event ErrorHandler OnInternalError;

    private static string GoodFilePath = Directory.GetCurrentDirectory() + @"\GoodFiles\";
    private static string ErrorFilePath = Directory.GetCurrentDirectory() + @"\ExceptionFiles\";
    private static string CSVPath = Directory.GetCurrentDirectory() + @"\CSVFiles\";


    public static void ParserLogFiles()
    {
        try
        {
            if (!Directory.Exists(GoodFilePath))
            {
                Directory.CreateDirectory(GoodFilePath);
            }
            if (!Directory.Exists(ErrorFilePath))
            {
                Directory.CreateDirectory(ErrorFilePath);
            }
            if (!Directory.Exists(CSVPath))
            {
                Directory.CreateDirectory(CSVPath);
            }

            var logFilePath = Directory.GetCurrentDirectory() + @"\LogFiles";
            if (!Directory.Exists(logFilePath))
            {
                Directory.CreateDirectory(logFilePath);
            }

            OnStatusReport.Invoke("Prepare Dummy log files");
            GenerateDummyLogFiles.GenerateLogFiles(logFilePath, 200);
            var files = Directory.GetFiles(logFilePath, "*.log");
            OnFileAmount.Invoke(files.Length);
            OnStatusReport.Invoke("Total [" + files.Length.ToString() + "] files are ready");
            Thread.Sleep(2000);
            ParallelExecution(files);
        }
        catch (Exception ex)
        {
            OnInternalError.Invoke(ex);
        }
    }

    private static void ParallelExecution(string[] files)
    {
        try
        {
            Task.Run(() =>
                    {
                        Stopwatch stopwatch = Stopwatch.StartNew();
                        Parallel.ForEach(files, new ParallelOptions
                        {
                            MaxDegreeOfParallelism = Environment.ProcessorCount
                        }, currentFile =>
                        {
                            var ExtractedResult = FileContentParser.ExtractFileContent(ReadFileContent(currentFile));
                            StoreResultToCSVFile(ExtractedResult);
                            MoveFileToErrorFolder(currentFile);
                            OnFileCompleted.Invoke(string.Format("The file [{0}] is done", currentFile));
                            OnFileAmountChanged.Invoke(1);
                        });
                        OnFileCompleted.Invoke("Parallel Execution Time: " + stopwatch.ElapsedMilliseconds);
                        stopwatch.Stop();
                    });
        }
        catch (Exception e)
        {
            WriteDataToLocalFiles(e.Message);
            OnInternalError.Invoke(e);
        }

    }

    private static string ReadFileContent(string sourceFile)
    {
        string fileContent = string.Empty;
        try
        {
            using (StreamReader sr = new StreamReader(sourceFile, Encoding.Unicode))
            {
                fileContent = sr.ReadToEnd();
            }
        }
        catch (Exception e)
        {
            WriteDataToLocalFiles(e.Message);
            OnInternalError.Invoke(e);
        }
        return fileContent;
    }

    private static void MoveFileToErrorFolder(string sourceFile)
    {
        try
        {
            string fileName = "_Err_" + Path.GetFileName(sourceFile);
            File.Move(sourceFile, ErrorFilePath + fileName);
        }
        catch (Exception e)
        {
            WriteDataToLocalFiles(e.Message);
            OnInternalError.Invoke(e);
        }
    }
    private static void MoveFileToGoodFolder(string sourceFile)
    {
        try
        {
            File.Move(sourceFile, GoodFilePath);
        }
        catch (Exception e)
        {
            WriteDataToLocalFiles(e.Message);
            OnInternalError.Invoke(e);
        }
    }

    public static void WriteDataToLocalFiles(string strData)
    {
        try
        {
            var path = Directory.GetCurrentDirectory();
            string folder = path + @"\ApplicationLogs\";
            if (!Directory.Exists(folder))
            {
                Directory.CreateDirectory(folder);
            }
            string fileName = folder + string.Format("{0:yyyyMMddhh}", DateTime.Now) + ".log";

            using (var fs = new FileStream(fileName, FileMode.OpenOrCreate, FileSystemRights.AppendData, FileShare.ReadWrite, 4096, FileOptions.Asynchronous))
            {
                using (var wr = new StreamWriter(fs))
                {
                    wr.AutoFlush = true;
                    wr.WriteLine(string.Format("{0:yyyy-MM-dd HH:mm:ss.ffff}", DateTime.Now));
                    wr.WriteLine($"{strData}\r");
                }
            }
        }
        catch (Exception ex)
        {
            OnInternalError.Invoke(ex);
        }
    }

    public static void StoreResultToCSVFile(string strData)
    {
        try
        {
            string fileName = CSVPath + string.Format("{0:yyyyMMddhh}", DateTime.Now) + ".csv";

            using (var fs = new FileStream(fileName, FileMode.OpenOrCreate, FileSystemRights.AppendData, FileShare.ReadWrite, 4096, FileOptions.Asynchronous))
            {
                using (var wr = new StreamWriter(fs))
                {
                    wr.AutoFlush = true;
                    wr.WriteLine($"{strData}\r");
                }
            }
        }
        catch (Exception ex)
        {
            WriteDataToLocalFiles(ex.Message);
            OnInternalError.Invoke(ex);
        }
    }

    public static class GenerateDummyLogFiles
    {
        public static void GenerateLogFiles(string destPath, int Qty)
        {
            try
            {
                //Task.Run(() =>
                //{
                Parallel.For(0, Qty, x =>
            {
                string destFile = destPath + "\\" + string.Format("{0:yyyyMMddHHmmssffff}", DateTime.Now) + "_" + x.ToString() + ".log";
                using (FileStream fs = File.Create(destFile))
                {
                    AddText(fs, "This is some text");
                    AddText(fs, "This is some more text,");
                    AddText(fs, "\r\nand this is on a new line");
                    AddText(fs, "\r\n\r\nThe following is a subset of characters:\r\n");

                    for (int i = 1; i < 120; i++)
                    {
                        AddText(fs, Convert.ToChar(i).ToString());
                    }
                    // OnStatusReport.Invoke("File [" + destFile + "] is created.");
                    //OnFileAmountChanged.Invoke(1);
                }
            });
                //});
            }
            catch (Exception ex)
            {
                WriteDataToLocalFiles(ex.Message);
                OnInternalError.Invoke(ex);
            }
        }
    }

    private static void AddText(FileStream fs, string value)
    {
        byte[] info = new UTF8Encoding(true).GetBytes(value);
        fs.Write(info, 0, info.Length);
    }
}

public static class FileContentParser
{
    public static string ExtractFileContent(string fileContent)
    {
        return "a,b,c,d";
    }
}
